master/cmd/stream-gen/main.go

```diff
@@ -0,0 +1,526 @@
+	b.Writef("\n")
+	b.Writef("    def __eq__(self, other: object) -> bool:\n")
+	b.Writef("        return isinstance(other, type(self)) and self.to_json() == other.to_json()\n")
+
+	allowedArgs := map[string]bool{
+		"delete_msg": true,
+		"source":     true,
+	}
+	requiredArgs := []string{"source"}
+
+	for _, s := range streamables {
+		// verify args
+		for k, v := range s.Args {
+			if !allowedArgs[k] {
+				fmt.Fprintf(os.Stderr, "unrecognized arg %q (%v=%v) @ %v\n", k, k, v, s.Position)
+				os.Exit(1)
+			}
+		}
+		for _, k := range requiredArgs {
+			if _, ok := s.Args[k]; !ok {
+				fmt.Fprintf(os.Stderr, "missing required arg %q @ %v\n", k, s.Position)
+				os.Exit(1)
+			}
+		}
+
+		source := s.Args["source"]
+
+		switch source {
+		case "server":
+			// Generate a subclass of a ServerMsg, all fields are always filled.
+			b.Writef("\n\n")
+			b.Writef("class %v(ServerMsg):\n", s.Name)
+			b.Writef("    def __init__(\n")
+			b.Writef("        self,\n")
+			for _, f := range s.Fields {
+				anno, err := typeAnno(f)
+				if err != nil {
+					return nil, fmt.Errorf("struct %v, field %v: %v", s.Name, f.Name, err)
+				}
+				b.Writef("        %v: %q,\n", f.JSONTag, anno)
+			}
+			b.Writef("    ) -> None:\n")
+			for _, f := range s.Fields {
+				b.Writef("        self.%v = %v\n", f.JSONTag, f.JSONTag)
+			}
+			if deleter := s.Args["delete_msg"]; deleter != "" {
+				// Also generate a Delete message
+				b.Writef("\n\n")
+				b.Writef("class %v(DeleteMsg):\n", deleter)
+				b.Writef("    pass\n")
+			}
+		case "client":
+			// Generate a subclass of a ClientMsg, all fields are always optional.
+			b.Writef("\n\n")
+			b.Writef("class %v(ClientMsg):\n", s.Name)
+			b.Writef("    def __init__(\n")
+			b.Writef("        self,\n")
+			for _, f := range s.Fields {
+				anno, err := typeAnno(f)
+				if err != nil {
+					return nil, fmt.Errorf("struct %v, field %v: %v", s.Name, f.Name, err)
+				}
+				b.Writef("        %v: %q = None,\n", f.JSONTag, optional(anno))
+			}
+			b.Writef("    ) -> None:\n")
+			for _, f := range s.Fields {
+				b.Writef("        self.%v = %v\n", f.JSONTag, f.JSONTag)
+			}
+		default:
+			fmt.Fprintf(os.Stderr, "invalid 'source' value %q @ %v\n", source, s.Position)
+			os.Exit(1)
+		}
+	}
+	return []byte(b.String()), nil
+}
+
+func printHelp(output io.Writer) {
+	fmt.Fprintf(
+		output,
+		`stream-gen generates bindings for determined streaming updates.
+
+usage: stream-gen IN.GO... --python [--output OUTPUT] [--stamp STAMP]
+
+All structs in the input files IN.GO... which contain special 'determined:stream-gen' comments will
+be included in the generated output.
+
+Presently the only output language is --python.
+
+Output will be written to stdout, or a location specified by --output.  The OUTPUT will only be
+overwritten if it would be modified.
+
+If --stamp is provided, the STAMP file will be touched after a successful run, which is useful for
+integration with build systems.
```

Comment:
I'm curious how the stamp file is used for integration with build systems. Is there an example in the code base that I can take a look at?

Reply:
you know, looking at this... there isn't a good reason for a `--stamp` output with how this function behaves.  I'll just remove this.